src/sage/rings/polynomial/polydict.pyx

```diff
@@ -42,131 +43,238 @@ from cpython.dict cimport *
             PolyDict with representation {(0, 0): 0.?}
+
+        TESTS::
+
+            sage: from sage.rings.polynomial.polydict import PolyDict
+            sage: f = PolyDict({(2, 3):2, (1, 2):3, (2, 1):4})
+            sage: len(f)
+            3
+            sage: f = PolyDict({}, zero=3, force_int_exponents=True, force_etuples=True)
+            doctest:warning
+            ...
+            DeprecationWarning: the arguments "zero", "forced_int_exponents"
+            and "forced_etuples" of PolyDict constructor are deprecated
+            See https://github.com/sagemath/sage/issues/34000 for details.
+            sage: f = PolyDict({}, remove_zero=False)
+            doctest:warning
+            ...
+            DeprecationWarning: the argument "remove_zero" of PolyDict
+            constructor is deprecated; call the method remove_zeros
+            See https://github.com/sagemath/sage/issues/34000 for details.
         """
+        if zero is not None or force_int_exponents is not None or force_etuples is not None:
+            from sage.misc.superseded import deprecation
+            deprecation(34000, 'the arguments "zero", "forced_int_exponents" and "forced_etuples" of PolyDict constructor are deprecated')
+
         cdef dict v
 
-        if not isinstance(pdict, dict):
-            if isinstance(pdict, list):
-                v = {}
-                L = <list> pdict
-                for w in L:
-                    if w[0] != 0:
-                        v[ETuple(w[1])] = w[0]
-                remove_zero = False
-                pdict = v
-            else:
-                raise TypeError("pdict must be a list")
-
-        if isinstance(pdict, dict) and force_etuples is True:
-            v = pdict
-            pdict = {}
-            for k, val in v.iteritems():
-                pdict[ETuple(k)] = val
-
-        if force_int_exponents:
-            new_pdict = {}
-            if remove_zero and zero is not None:
-                for k, c in pdict.iteritems():
-                    if not c == zero:
-                        new_pdict[ETuple([int(i) for i in k])] = c
-            else:
-                for k, c in pdict.iteritems():
-                    new_pdict[ETuple([int(i) for i in k])] = c
-            pdict = new_pdict
+        if isinstance(pdict, (tuple, list)):
+            v = {}
+            for w in pdict:
+                v[ETuple(w[1])] = w[0]
+        elif isinstance(pdict, dict):
+            v = <dict> pdict.copy()
         else:
-            if remove_zero and zero is not None:
-                for k in list(pdict):
-                    if pdict[k] == zero:
-                        del pdict[k]
-        self.__repn = <dict> pdict
-        self.__zero = zero
+            raise TypeError("pdict must be a dict or a list")
+
+        for k in list(v):
+            if type(k) is not ETuple:
+                val = v[k]
+                del v[k]
+                v[ETuple(k)] = val
```

Comment:
This is generally slower than what it was before as now it explicitly checks everything everytime for the keys being an `ETuple`. I am -1 on removing the argument that disabled this check (which we can assume for internal code that things are good).

Reply:
What do you refer to with _the argument that disabled this check_?

More importantly the check `type(X) is ETuple` is cythonized as
```
(((PyObject *)Py_TYPE(__pyx_t_1)) == ((PyObject *)__pyx_ptype_4sage_5rings_10polynomial_8polydict_ETuple))
```
You can notice that it converts as a call to `Py_TYPE` (which is an extremely fast inline function) and a single C pointer comparison. In particular, the conversion `ETuple(k)` in the above code is much slower than the type check.

Though, I do not particularly like the `list(v)` which creates a list for possibly no purpose. I will propose an alternative version.

Reply:
See https://github.com/sagemath/sage/pull/35174/commits/3b1b364db624c83e7236e52fa94010e4a542994b

Reply:
It still is a cost, whereas before, if you set `force_etuple=False`, then _no_ checks were run. This is code we should really pay attention to the speed aspects.

Reply:
Can you suggest an example where a speed regression can be noticed?

Reply:
Either a reasonably large polynomial or doing something in a very frequently called loop.

Reply:
Perhaps something like this would be sufficient:
```python
sage: R.<x,y> = PolynomialRing(CyclotomicField(5)['q'].fraction_field())
sage: p = x*y
sage: %timeit p._MPolynomial_element__element.polynomial_coefficient([1,1])
```

Reply:
As a first comparison (data with ~50000 terms)
```python
sage: from sage.rings.polynomial.polydict import ETuple, PolyDict
sage: import itertools
sage: data = {ETuple(t): 1 for t in itertools.product([1r, 2r, 3r, 4r], repeat=8)}
```
gives with the old code
```python
sage: %timeit p = PolyDict(data)
85.3 ms ± 747 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
sage: PolyDict??
sage: %timeit p = PolyDict(data, force_etuples=False)
65.4 ms ± 398 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```
while with the new code
```
sage: %timeit p = PolyDict(data)
1.77 ms ± 9.62 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```

Reply:
Is it sufficient?

Reply:
That's not a fair comparison. You have made other optimizations. You would need to implement a parameter for the constructor that skips the `ETuples` checks.

Reply:
https://github.com/sagemath/sage/pull/35174/commits/dfd6609d86758031cad774edeb8b078cc2cdeba3 introduces a `check` argument.

Reply:
Gets roughly x1.7 faster on the big example above.